core/arch/arm/kernel/generic_entry_a32.S

```diff
@@ -425,6 +445,9 @@ UNWIND(	.cantunwind)
 
 	bl	plat_cpu_reset_late
 
+    /* this core just comes out of reset, discard undefined l1 contents */
+	bl	arm_cl1_d_invbysetway
```

Comment:
> This is safe because its "cache invalidate" on L1, other cores would not
respond to this op.

ok. thanks for the details.
looks clear.


> And I wonder why this is not required (...)

on armv8, cache were cleaned by el3.
on armv7, it is likely that for those legacy platforms an earlier
bootloader invalidates caches.


> I'll put it inside CFG_TEE_VXWORKS_SUPPORT to reflect this is due to
VxWorks's special boot scenario.

since introducing CFG_TEE_VXWORKS_SUPPORT , it could be nice to have this
info inside optee.
Could you elaborate some make command using
$(call cfg-depends-one/all,... to detect the dependency, and integrate it
to mk/config.mk ?


On 12 August 2016 at 10:04, yanyan-wrs <notifications@github.com> wrote:

> In core/arch/arm/kernel/generic_entry_a32.S
> <https://github.com/OP-TEE/optee_os/pull/936#discussion_r74554282>:
>
> > @@ -425,6 +445,9 @@ UNWIND(	.cantunwind)
> >
> >  	bl	plat_cpu_reset_late
> >
> > +    /* this core just comes out of reset, discard undefined l1 contents */
> > +	bl	arm_cl1_d_invbysetway
>
> This is safe because its "cache invalidate" on L1, other cores would not
> respond to this op. Besides, the MMU and cache for secondary core is not
> enabled yet at this moment, so I believe there will not be any coherency
> broadcast even if the SMP bit is enable.
>
> Anyway, my testing on i.MX6 (Cortex-A9) shows this operation is required
> for secondary cores to boot up. ARM specs also mentioned that when CPU
> comes out of reset, the cache content is UNDEFINED, so an all set-way
> invalidation is necessary.
>
> And I wonder why this is not required before. Maybe this again is because
> the OP-TEE is targeted for Firmware; so it is possible that Firmware
> cleaned the cache before putting secondary cores into OP-TEE; however, for
> VxWorks's usage, the secondary core is released from reset and is directly
> put into OP-TEE, so the cache should be handled somehow here.
>
> Anyway, I admit globally changing this is inconsiderate, I'll put it
> inside CFG_TEE_VXWORKS_SUPPORT to reflect this is due to VxWorks's special
> boot scenario.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OP-TEE/optee_os/pull/936/files/94967ff72e6c157bd08e0134b1e82a5792cfd333#r74554282>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ATEHJsT1eLSULwkjlwh4RAzH6Ngzz6Zmks5qfCkjgaJpZM4JTw0v>
> .
>